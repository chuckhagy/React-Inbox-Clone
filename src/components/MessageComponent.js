import React from 'react';
import preloader from '../preloader.gif'
let moment = require('moment');

export default class MessageComponent extends React.Component{
  state = {
    bodyVisibile: false,
  }

  handleReadClick = event => {
    event.preventDefault();
    if(!this.state.bodyVisible) this.setState({bodyVisible: true})
    if(this.state.bodyVisible) this.setState({bodyVisible: false})
    this.props.onMarkAsReadMessage(this.props.message.id);
  }

    handleStarClick = event => {
    event.preventDefault();
    if(this.props.message.starred) this.props.onUnstarMessage(this.props.message.id);
    else this.props.onStarMessage(this.props.message.id);
  }

    handleSelectClick = event => {
    if(this.props.selected) this.props.onDeselectMessage(this.props.message.id);
    else this.props.onSelectMessage(this.props.message.id);
  }

  render(){
    if (!this.props.message) return <h5>"No Matches"</h5>;

    let readClass = '';
    if (!this.props.message.read) {
      readClass = 'unread';
    } else {
      readClass = 'read';
    }

    let starClass = '';
    if (this.props.message.starred) {
      starClass = 'fa-star';
    } else {
      starClass = 'fa-star-o';
    }

    let selectedClass = '';
    if (this.props.selected) {
      selectedClass = 'selected';
    } else {
      selectedClass = '';
    }
    
    let fullDate = this.props.date;
    let diff = new Date(fullDate) - Date.now();
    let date = null;
    if (fullDate){
      if(diff < -86400000){
        date = moment(fullDate).format("MMM Do, YYYY")
      }
      else if(diff < -3600000){
        date = moment(fullDate).format("h:mm a")
      } else{
        date = moment(fullDate).fromNow()
      }
    }

  return (
    <div>

    <div className={`row message ${readClass} ${selectedClass}`}>

      <div className="col-xs-1">
        <div className="row">
          <div className="col-xs-2">
            <input type="checkbox"  checked={!!this.props.selected} onClick={this.handleSelectClick}/>
          </div>
          <div className="col-xs-2">
            {this.props.starLoading === this.props.message.id || this.props.unstarLoading === this.props.message.id ? <img alt='loader' src={preloader} style={{width: '16px'}}/> : <i className={`star fa ${starClass}`} onClick={this.handleStarClick} />}
          </div>
        </div>
      </div>
      <div className="col-xs-11" onClick={this.handleReadClick} >
        {this.props.message.labels.map((label, index) => <span className="label label-warning" key={index}>{label}</span>)}
        <a href="." >
          {this.props.message.subject}
        </a>
        <span style={{position: "absolute", right: 15}}>{date}</span>
      </div>
    </div>
    {this.state.bodyVisible ? <div className="message-body">{this.props.message.body}</div> : null}
  </div>
  )};
}
