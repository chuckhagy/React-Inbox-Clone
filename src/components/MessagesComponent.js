import React from 'react';
import MessageComponent from './MessageComponent';


export default function MessagesComponent({
  messages,
  selectedMessageIds,
  onMarkAsReadMessage,
  onStarMessage,
  onUnstarMessage,
  onSelectMessage,
  onDeselectMessage,
  starLoading,
  unstarLoading
}){
  if (!messages) return <h5>"No Messages"</h5>;
  return(
    <div>
      {messages.map(message =>
        <MessageComponent
          message={message}
          key={message.id}
          selected={selectedMessageIds.includes(message.id)}
          onMarkAsReadMessage={onMarkAsReadMessage}
          onStarMessage={onStarMessage}
          onUnstarMessage={onUnstarMessage}
          onSelectMessage={onSelectMessage}
          onDeselectMessage={onDeselectMessage}
          selectedMessageIds={selectedMessageIds}
          date={message.date}
          starLoading={starLoading}
          unstarLoading={unstarLoading}


      />)}
    </div>
  )
}
