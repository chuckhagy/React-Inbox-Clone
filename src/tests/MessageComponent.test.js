import React from 'react';
import { shallow, mount } from 'enzyme';
import MessageComponent from '../components/MessageComponent'
import MessagesComponent from '../components/MessagesComponent'

const message = {
    subject: 'Hi',
    body: "wazzup",
    starred: true,
    read: true,
    labels: ["a", "b"]
  }
  
  let selectedMessageIds = [1, 2]
  
  let messages = [
  {
  "id": 1,
  "subject": "Free Trial! MeetChuck.com This week only!",
  "read": false,
  "starred": true,
  "labels": ["hot deals", "personal"]
  },
  {
  "id": 2,
  "subject": "connecting the system won't do anything, we need to input the mobile AI panel!",
  "read": false,
  "starred": false,
  "selected": true,
  "labels": []
  },
  {
  "id": 3,
  "subject": "Use the 1080p HTTP feed, then you can parse the cross-platform hard drive!",
  "read": false,
  "starred": true,
  "labels": ["dev"]
  }
  ]
  
const selected = true;
  
describe('shallow testing portion', ()  => {
  let shallowWrapper = shallow(<MessageComponent message={message} selected={selected} />);
  it('should appear with subject', () => {
    expect(shallowWrapper.find('a')).toHaveLength(1);
  });
  it('should appear READ when data says so and vice versa', () => {
    if(message.read) expect(shallowWrapper.find('.read')).toHaveLength(1);
    else expect(shallowWrapper.find('.read')).toHaveLength(0);
  });
  it('should appear STARRED when data says so and vice versa', () => {
    if(message.star) expect(shallowWrapper.find('.star')).toHaveLength(1);
    else expect(shallowWrapper.find('.star')).toHaveLength(0);
  });
  it('should appear CHECKED when selected prop passed down', () => {
    if(selected) expect(shallowWrapper.find('input').prop('checked')).toBe(true);
    else expect(shallowWrapper.find('input').prop('checked')).toBe(false);
  })
  it('should fire onMarkAsReadMessage when the subject is clicked', () => {
    const onMarkAsReadMessage = jest.fn();
    mount(<MessageComponent message={message} selected={selected} onMarkAsReadMessage={onMarkAsReadMessage} />).find('a').simulate('click');
    expect(onMarkAsReadMessage).toHaveBeenCalled()
  });
  it('should fire onSelectMessage when the subject is clicked - and Vice Versa', () => {
    const onSelectMessage = jest.fn();
    const onDeselectMessage = jest.fn();
    mount(<MessageComponent message={message} selected={selected} onSelectMessage={onSelectMessage} onDeselectMessage={onDeselectMessage} />).find('input').simulate('click');
    if(selected)expect(onDeselectMessage).toHaveBeenCalled()
    else expect(onSelectMessage).toHaveBeenCalled()
  });
  // it('should fire onStarMessage when the subject is clicked - and Vice Versa', () => {
  //   const onStarMessage = jest.fn();
  //   const onUnstarMessage = jest.fn();
  //   mount(<MessageComponent message={message} selected={selected} onStarMessage={onStarMessage} onUnstarMessage={onUnstarMessage} />).find('.star').simulate('click');
  //   if(message.starred)expect(onUnstarMessage).toHaveBeenCalled()
  //   else expect(onStarMessage).toHaveBeenCalled()
  // });
});
  

describe('Messages Component Tests', ()  => {
  it('should have correct number of messages displayed', () => {
    let count = mount(<MessagesComponent messages={messages} selectedMessageIds={selectedMessageIds} />).find('.message').length;
    expect(count).toEqual(messages.length)
  });
  it('should trigger callbacks OnSelectMessage and OnDeSelectMessage', () => {
    const onSelectMessage = jest.fn();
    const onDeselectMessage = jest.fn();
    let myThing = mount(<MessagesComponent
      messages={messages}
      selectedMessageIds={selectedMessageIds}
      onSelectMessage={onSelectMessage}
      onDeselectMessage={onDeselectMessage}

    />).find('input').at(0);
    
    myThing.simulate('click');
    if(myThing.prop('checked')) expect(onDeselectMessage).toHaveBeenCalled()
    else expect(onSelectMessage).toHaveBeenCalled();
    });
    
  it('should trigger callbacks onMarkAsReadMessage', () => {
    const onMarkAsReadMessage = jest.fn();
    let myThing = mount(<MessagesComponent
      messages={messages}
      selectedMessageIds={selectedMessageIds}
      onMarkAsReadMessage={onMarkAsReadMessage}

    />).find('a').at(0);
    
    myThing.simulate('click');
    if(!messages[0].read) expect(onMarkAsReadMessage).toHaveBeenCalled()
    });


});