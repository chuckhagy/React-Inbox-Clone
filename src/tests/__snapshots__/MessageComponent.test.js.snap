// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shallow testing portion should match shallow wrapper 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div>
    <div
        className="row message read selected"
    >
        <div
            className="col-xs-1"
        >
            <div
                className="row"
            >
                <div
                    className="col-xs-2"
                >
                    <input
                        checked={true}
                        onClick={[Function]}
                        type="checkbox"
                    />
                </div>
                <div
                    className="col-xs-2"
                >
                    <img
                        alt="loader"
                        src="preloader.gif"
                        style={
                            Object {
                                "width": "16px",
                              }
                        }
                    />
                </div>
            </div>
        </div>
        <div
            className="col-xs-11"
            onClick={[Function]}
        >
            <span
                className="label label-warning"
            >
                a
            </span>
            <span
                className="label label-warning"
            >
                b
            </span>
            <a
                href="."
            >
                Hi
            </a>
            <span
                style={
                    Object {
                        "position": "absolute",
                        "right": 15,
                      }
                }
            />
        </div>
    </div>
</div>,
  "nodes": Array [
    <div>
      <div
            className="row message read selected"
      >
            <div
                  className="col-xs-1"
            >
                  <div
                        className="row"
                  >
                        <div
                              className="col-xs-2"
                        >
                              <input
                                    checked={true}
                                    onClick={[Function]}
                                    type="checkbox"
                              />
                        </div>
                        <div
                              className="col-xs-2"
                        >
                              <img
                                    alt="loader"
                                    src="preloader.gif"
                                    style={
                                          Object {
                                                "width": "16px",
                                              }
                                    }
                              />
                        </div>
                  </div>
            </div>
            <div
                  className="col-xs-11"
                  onClick={[Function]}
            >
                  <span
                        className="label label-warning"
                  >
                        a
                  </span>
                  <span
                        className="label label-warning"
                  >
                        b
                  </span>
                  <a
                        href="."
                  >
                        Hi
                  </a>
                  <span
                        style={
                              Object {
                                    "position": "absolute",
                                    "right": 15,
                                  }
                        }
                  />
            </div>
      </div>
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <MessageComponent
        message={
                Object {
                        "body": "wazzup",
                        "labels": Array [
                          "a",
                          "b",
                        ],
                        "read": true,
                        "starred": true,
                        "subject": "Hi",
                      }
        }
        selected={true}
/>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": MessageComponent {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "handleReadClick": [Function],
        "handleSelectClick": [Function],
        "handleStarClick": [Function],
        "props": Object {
          "message": Object {
            "body": "wazzup",
            "labels": Array [
              "a",
              "b",
            ],
            "read": true,
            "starred": true,
            "subject": "Hi",
          },
          "selected": true,
        },
        "refs": Object {},
        "state": Object {
          "bodyVisibile": false,
        },
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div>
          <div
                    className="row message read selected"
          >
                    <div
                              className="col-xs-1"
                    >
                              <div
                                        className="row"
                              >
                                        <div
                                                  className="col-xs-2"
                                        >
                                                  <input
                                                            checked={true}
                                                            onClick={[Function]}
                                                            type="checkbox"
                                                  />
                                        </div>
                                        <div
                                                  className="col-xs-2"
                                        >
                                                  <img
                                                            alt="loader"
                                                            src="preloader.gif"
                                                            style={
                                                                      Object {
                                                                                "width": "16px",
                                                                              }
                                                            }
                                                  />
                                        </div>
                              </div>
                    </div>
                    <div
                              className="col-xs-11"
                              onClick={[Function]}
                    >
                              <span
                                        className="label label-warning"
                              >
                                        a
                              </span>
                              <span
                                        className="label label-warning"
                              >
                                        b
                              </span>
                              <a
                                        href="."
                              >
                                        Hi
                              </a>
                              <span
                                        style={
                                                  Object {
                                                            "position": "absolute",
                                                            "right": 15,
                                                          }
                                        }
                              />
                    </div>
          </div>
</div>,
        "_debugID": 2,
        "_renderedOutput": <div>
          <div
                    className="row message read selected"
          >
                    <div
                              className="col-xs-1"
                    >
                              <div
                                        className="row"
                              >
                                        <div
                                                  className="col-xs-2"
                                        >
                                                  <input
                                                            checked={true}
                                                            onClick={[Function]}
                                                            type="checkbox"
                                                  />
                                        </div>
                                        <div
                                                  className="col-xs-2"
                                        >
                                                  <img
                                                            alt="loader"
                                                            src="preloader.gif"
                                                            style={
                                                                      Object {
                                                                                "width": "16px",
                                                                              }
                                                            }
                                                  />
                                        </div>
                              </div>
                    </div>
                    <div
                              className="col-xs-11"
                              onClick={[Function]}
                    >
                              <span
                                        className="label label-warning"
                              >
                                        a
                              </span>
                              <span
                                        className="label label-warning"
                              >
                                        b
                              </span>
                              <a
                                        href="."
                              >
                                        Hi
                              </a>
                              <span
                                        style={
                                                  Object {
                                                            "position": "absolute",
                                                            "right": 15,
                                                          }
                                        }
                              />
                    </div>
          </div>
</div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <MessageComponent
    message={
        Object {
            "body": "wazzup",
            "labels": Array [
              "a",
              "b",
            ],
            "read": true,
            "starred": true,
            "subject": "Hi",
          }
    }
    selected={true}
/>,
}
`;
